insights/parsers/audit_log.py

```diff
@@ -0,0 +1,164 @@
+            'unparsed': 'avc:  denied  { create } for',
+            'msg_ID': '32794',
+            'pid': '27960',
+            'raw_message': 'type=AVC msg=audit(1506487181.009:32794): avc:  denied  { create } for  pid=27960 comm="mongod" scontext=system_u:system_r:mongod_t:s0 tcontext=system_u:system_r:mongod_t:s0 tclass=unix_dgram_socket',
+            'comm': 'mongod',
+            'scontext': 'system_u:system_r:mongod_t:s0',
+            'tclass': 'unix_dgram_socket',
+            'type': 'AVC',
+            'tcontext': 'system_u:system_r:mongod_t:s0'
+        }]
+        >>> assert len(list(log.get_after(timestamp=date.fromtimestamp(1506047401.407)))) == 3
+    """
+    def _parse_line(self, line):
+        """
+        Parse a log line into a info dictionary.
+
+        Parsing logic:
+
+            * First, use `shlex.split` to split by empty.
+            * Next, assert the first two key-value pair is 'type' and 'msg'.
+            * Next, parse the remained string reversly to get key-value pair data as more as possible.
+            * The left unparsed string will be stored at "unparsed".
+            * At last, wrap all key-value pairs into a dict called info.
+
+        Parameters:
+            line(string): Log line to parse
+
+        Yields:
+            Dictionary corresponding to the data in line split into
+            functional elements:
+
+            * **raw_message** - the original unparsed line
+            * **type** - the log type
+            * **timestamp** - the timestamp of the log line (as a string)
+            * **msg_ID** - the serial number behind the timestamp
+            * **unparsed** - the unparsed part in log, may or maynot exist
+
+            If the line is not regular, for some reason, then as many fields as
+            possible are pulled from the line.
+        """
+        info = {}
+        linesp = shlex.split(line)
+
+        if (len(linesp) < 2 or
+                not (linesp[0] and linesp[0].startswith('type=')) or
+                not (linesp[1] and linesp[1].startswith('msg=audit('))):
+            return
+
+        info['type'] = linesp[0].lstrip('type=')
+
+        timestamp_id = linesp[1].lstrip('msg=audit(').rstrip('):')
+        timestamp_id_sp = timestamp_id.split(':')
+        if len(timestamp_id_sp) != 2:
+            return
+        info['timestamp'] = timestamp_id_sp[0]
+        info['msg_ID'] = timestamp_id_sp[1]
+
+        for index in xrange(len(linesp) - 1, 1, -1):
+            itemsp = linesp[index].split('=', 1)
+            if len(itemsp) < 2:
+                unparsed1 = (line.rsplit(linesp[index + 1])[0]
+                             if index + 1 < len(linesp)
+                             else line)
+                unparsed2 = unparsed1.split(linesp[1])[-1]
+                info["unparsed"] = unparsed2.strip()
+                break
+            info[itemsp[0]] = itemsp[1]
+
+        info['raw_message'] = line
+        return info
+
+    def get(self, s):
```

Comment:
Then this `get` can be removed. Since once a line contains the keyword `s`, it's OK to return the `raw_message` even it cannot be parsed.